### PR TITLE
Refactor recognition of loads of multiramps

### PR DESCRIPTION
**Pull request by abadams**
This is a refactoring of some AMX matrix-load-recognition logic so that it can be shared with other targets that need to recognize the same basic patterns: multi-dimensional loads, and multi-dimensional loads of non-overlapping subtiles. Recognizing a multi-dimensional load is slightly more tricky that asking if a load index is a multiramp, because there may be shuffles and broadcasts *outside* the load too - shuffles and loads commute. The second feature is needed to partition one logical Halide allocation between multiple AMX tile registers (or in the future, WMMA fragments).

### CODE REVIEW

**Review by alexreinking — DISMISSED** [review]

**Comment by abadams** [comment]
I had to redo shuffles-of-multiramps to support more than the 1d case for future work. Please take another look at multiramp.cpp

File: src/MultiRamp.cpp [hunk 1]
```diff
@@ -34,6 +34,40 @@ void collapse_adjacent_dims(MultiRamp *m) {
...
+        y /= g;
+        if (x == 1 && ++i < a.size()) {
+            x = a[i];
+        }
+        if (y == 1 && ++j < b.size()) {
+            y = b[j];
+        }
+    }
+    return true;
+}
```
Comment (alexreinking) [inline src/MultiRamp.cpp:96]: I had to read this through a few times... the "coprime" language is misleading. For instance, `6` and `8` are not coprime, but are clearly different shapes. The invariant that must hold is that one of `x` or `y` divides the other. You should check that directly:

```c++
int g = std::min(x, y);
if (std::max(x, y) % g != 0) {
    return false;
}
```

I'm also suspicious of the handling of empty vectors. I think it breaks on `a=[]` and `b = [6]`. The function returns `true` and `refined = []`. Why doesn't it return `false`? This is one possible correction:

```c++
if (a.empty() || b.empty()) {
    return a.empty() && b.empty();
}
size_t i = 0, j = 0;
int x = a[0], y = b[0];
```

Please make these changes and add regression tests for empty vectors. Also update the doc-comment with a worked example illustrating what it does and correct the claim concerning "coprime".

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: src/MultiRamp.cpp (hunk 1 above)
Comment (abadams) [reply src/MultiRamp.cpp:96]: All the callers of common_refinement have already asserted the lane counts match. I'll just assert it here too. It should assert fail, not return false.

Comment (abadams) [reply src/MultiRamp.cpp:96]: With the assert that both vectors are the same size, I think the empty-vector case is trivially correct.

Comment (abadams) [reply src/MultiRamp.cpp:96]: update pushed

**Review by alexreinking — APPROVED** [review]
